File: src/adapters/claude_code/history_jsonl.ts
import { open } from 'node:fs/promises';
import { z } from 'zod';
import type { Repository } from '../../store/repository.js';
import { normalizeProjectPath } from '../../store/repository.js';
import type { Prompt } from '../../schema/types.js';

// Hard cap on the indexed body. Eight kilobytes is way past any realistic
// hand-typed prompt; anything longer is almost always a paste-by-mistake
// where the user pasted a file into the prompt without using the paste
// shortcut. We truncate so FTS5 doesn't try to index a multi-megabyte row.
const DISPLAY_CAP_BYTES = 8 * 1024;

// One line of ~/.claude/history.jsonl. The file is global (not per-session)
// and Claude Code appends to it on every prompt submission.
const HistoryLineSchema = z.object({
  display: z.string(),
  pastedContents: z.record(z.unknown()).optional().default({}),
  timestamp: z.number(),
  project: z.string(),
});

export interface HistoryIngestStats {
  inserted: number;
  skipped_empty: number;
  parse_errors: number;
  new_offset: number;
}

// Read ~/.claude/history.jsonl from the stored byte offset, parse new lines,
// and write them to the prompts table. Mirrors the byte-offset tail pattern
// used for session JSONL — partial last line is held back until the next
// tick. If the file has shrunk below the recorded offset (rotation /
// truncation), we reset to 0 and re-ingest from scratch.
//
// The repository is the only consumer of these rows: we don't return the
// inserted rows themselves, just stats that callers can log.
export async function ingestHistoryFile(
  filePath: string,
  repo: Repository,
): Promise<HistoryIngestStats> {
  const offsetKey = `history:${filePath}`;
  let fromOffset = repo.getFileOffset(offsetKey);

  const fh = await open(filePath, 'r');
  try {
    const stat = await fh.stat();
    const size = stat.size;

    // Rotation / truncation: file is smaller than the offset we last
    // recorded. Treat as fresh — re-ingest from the start. The prompts
    // table has no uniqueness constraint, so this can produce duplicates;
    // worth fixing only if it ever happens in practice.
    if (size < fromOffset) fromOffset = 0;

    if (size <= fromOffset) {
      return { inserted: 0, skipped_empty: 0, parse_errors: 0, new_offset: fromOffset };
    }

    const buf = Buffer.alloc(size - fromOffset);
    await fh.read(buf, 0, buf.length, fromOffset);
    const text = buf.toString('utf8');

    const lastNewline = text.lastIndexOf('\n');
    const consumable = lastNewline === -1 ? '' : text.slice(0, lastNewline + 1);
    const consumed_bytes = Buffer.byteLength(consumable, 'utf8');
    const new_offset = fromOffset + consumed_bytes;

    const rows: Omit<Prompt, 'id'>[] = [];
    let parse_errors = 0;
    let skipped_empty = 0;

    for (const line of consumable.split('\n')) {
      if (!line) continue;
      let obj: unknown;
      try {
        obj = JSON.parse(line);
      } catch {
        parse_errors++;
        continue;
      }
      const parsed = HistoryLineSchema.safeParse(obj);
      if (!parsed.success) {
        parse_errors++;
        continue;
      }
      const row = lineToPrompt(parsed.data);
      if (row === null) {
        skipped_empty++;
        continue;
      }
      rows.push(row);
    }

    repo.insertPrompts(rows);
    repo.setFileOffset(offsetKey, new_offset);

    return { inserted: rows.length, skipped_empty, parse_errors, new_offset };
  } finally {
    await fh.close();
  }
}

// Pure transformation: history.jsonl line → prompts row, or null if the
// line should be skipped (empty/whitespace-only display). Exposed for unit
// testing without touching the filesystem.
export function lineToPrompt(line: z.infer<typeof HistoryLineSchema>): Omit<Prompt, 'id'> | null {
  const trimmed = line.display.trim();
  if (trimmed === '') return null;

  // Cap by bytes, not chars, so unicode-heavy prompts don't blow past the
  // cap due to multi-byte characters. Truncation is suffix-marked so a
  // future reader can tell something was elided.
  let display = line.display;
  if (Buffer.byteLength(display, 'utf8') > DISPLAY_CAP_BYTES) {
    // Walk back from the cap to a safe utf-8 boundary by repeatedly
    // re-encoding the slice until it fits. Cheap because the cap is
    // small.
    let cap = DISPLAY_CAP_BYTES;
    while (cap > 0 && Buffer.byteLength(display.slice(0, cap), 'utf8') > DISPLAY_CAP_BYTES - 1) {
      cap--;
    }
    display = display.slice(0, cap) + '…';
  }

  // pasted_chars is a cheap upper bound on paste size — JSON-stringified
  // length of the whole pastedContents object. We never index the body.
  let pasted_chars = 0;
  try {
    pasted_chars = JSON.stringify(line.pastedContents ?? {}).length;
    if (pasted_chars <= 2) pasted_chars = 0; // "{}" → no paste
  } catch {
    pasted_chars = 0;
  }

  return {
    timestamp_ms: line.timestamp,
    project_path: normalizeProjectPath(line.project),
    display,
    pasted_chars,
    is_slash: trimmed.startsWith('/') ? 1 : 0,
  };
}

File: python/argus/adapters/claude_code/history_jsonl.py
"""Ingest ``~/.claude/history.jsonl`` — the global prompt history file.

Byte-offset tail like sessions. Partial last line is held back until
newline. If the file shrinks below the recorded offset (rotation /
truncation), we reset to 0 and re-ingest.
"""
from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING, Any

from pydantic import BaseModel, ConfigDict, Field, ValidationError

from ...schema.types import Prompt
from ...store.repository import normalize_project_path

if TYPE_CHECKING:
    from ...store.repository import Repository

# Hard cap on the indexed body. 8 KB is past any realistic hand-typed
# prompt; longer values are almost always a paste-by-mistake.
DISPLAY_CAP_BYTES = 8 * 1024
MAX_TICK_BYTES = 64 * 1024 * 1024


class HistoryLine(BaseModel):
    """One line of ~/.claude/history.jsonl."""

    model_config = ConfigDict(extra="allow")

    display: str
    pastedContents: dict[str, Any] = Field(default_factory=dict)
    timestamp: int
    project: str


class HistoryIngestStats(BaseModel):
    inserted: int
    skipped_empty: int
    parse_errors: int
    new_offset: int


def line_to_prompt(line: HistoryLine) -> Prompt | None:
    """Pure transformation: history line → Prompt row, or None if skipped."""
    trimmed = line.display.strip()
    if trimmed == "":
        return None

    display = line.display
    if len(display.encode("utf-8")) > DISPLAY_CAP_BYTES:
        # Walk back to a unicode-safe boundary by decoding with errors='ignore'.
        truncated = display.encode("utf-8")[: DISPLAY_CAP_BYTES - 1].decode(
            "utf-8", errors="ignore"
        )
        display = truncated + "…"

    try:
        # Compact separators match JS ``JSON.stringify`` byte count, so
        # the pasted_chars upper bound is identical across stacks.
        pasted_chars = len(json.dumps(line.pastedContents or {}, separators=(",", ":")))
        if pasted_chars <= 2:  # "{}"
            pasted_chars = 0
    except (TypeError, ValueError):
        pasted_chars = 0

    return Prompt(
        timestamp_ms=line.timestamp,
        project_path=normalize_project_path(line.project),
        display=display,
        pasted_chars=pasted_chars,
        is_slash=1 if trimmed.startswith("/") else 0,
    )


def ingest_history_file(
    file_path: Path, repo: "Repository"
) -> HistoryIngestStats:
    """Read ``file_path`` from the stored offset, append new prompts."""
    offset_key = f"history:{file_path}"
    from_offset = repo.get_file_offset(offset_key)

    with open(file_path, "rb") as fh:
        fh.seek(0, 2)
        size = fh.tell()

        # Rotation / truncation: file smaller than the offset → re-ingest.
        if size < from_offset:
            from_offset = 0

        if size <= from_offset:
            return HistoryIngestStats(
                inserted=0, skipped_empty=0, parse_errors=0, new_offset=from_offset
            )

        read_len = min(size - from_offset, MAX_TICK_BYTES)
        fh.seek(from_offset)
        raw = fh.read(read_len)

    text = raw.decode("utf-8", errors="replace")
    last_nl = text.rfind("\n")
    consumable = text[: last_nl + 1] if last_nl != -1 else ""
    consumed_bytes = len(consumable.encode("utf-8"))
    new_offset = from_offset + consumed_bytes

    rows: list[Prompt] = []
    parse_errors = 0
    skipped_empty = 0

    for line in consumable.split("\n"):
        if not line:
            continue
        try:
            obj = json.loads(line)
        except json.JSONDecodeError:
            parse_errors += 1
            continue
        try:
            parsed = HistoryLine.model_validate(obj)
        except ValidationError:
            parse_errors += 1
            continue
        row = line_to_prompt(parsed)
        if row is None:
            skipped_empty += 1
            continue
        rows.append(row)

    repo.insert_prompts(rows)
    repo.set_file_offset(offset_key, new_offset)

    return HistoryIngestStats(
        inserted=len(rows),
        skipped_empty=skipped_empty,
        parse_errors=parse_errors,
        new_offset=new_offset,
    )
